fix(migration-analyzer): correct misleading lock messaging and scoring

webjunkie opened the pull request:
Fix misleading "No locks" messaging in migration analyzer that contributed to production contention on Nov 4, 2025.

Migration 0900 (team_receive_org_level_activity_logs) triggered transient lock contention despite being labeled as
`✅ Safe - No locks`. The wording was technically wrong and misleading, "safe" in our analyzer has always meant *no table rewrite / metadata-only DDL*, not the complete absence of locks.  
However, the "No locks" phrase failed to communicate that **all `ALTER TABLE` statements briefly acquire an `ACCESS EXCLUSIVE` lock** even when the operation itself is safe and instantaneous.

This change clarifies the analyzer’s intent: such migrations are still safe to run in production when engineers understand the brief locking behavior and apply normal operational discipline (avoiding heavy traffic windows, monitoring long transactions, etc.).  
In other words, the tool should highlight *engineering judgment*, not imply magic immunity from Postgres semantics.

**Changes**

**Risk Level Categorization:**
- Changed RiskLevel ranges to accurately reflect lock impact:
  - SAFE (score 0): Only truly no-contention operations (CreateModel, concurrent ops)
  - NEEDS_REVIEW (score 1-3): Operations requiring locks or review
  - BLOCKED (score 4-5): Unchanged
- AddField nullable operations now appear in yellow NEEDS_REVIEW section instead of green SAFE section

**Scoring:**
- Changed AddField nullable from score 0 to score 1 (now NEEDS_REVIEW level)
- Score 0 reserved for truly contention-free operations (CreateModel, concurrent index ops)
- Score 1 covers operations that acquire brief ACCESS EXCLUSIVE locks (AddField on existing tables)

**Messaging:**
- Removed misleading "No locks" text from SAFE description
- SAFE: "No contention risk, backwards compatible" (accurate for score 0 only)
- NEEDS_REVIEW: "Requires brief lock, review for high-traffic tables" (accurate for score 1)
- Added guidance text for AddField operations about high-traffic table considerations

webjunkie committed:
fix(migration-analyzer): correct misleading lock messaging and scoring

Changed AddField nullable scoring from 0 to 1 and updated messaging to accurately reflect that ALTER TABLE operations acquire ACCESS EXCLUSIVE locks, even for safe operations like adding nullable fields.

Background:
Migration 0900 (team_receive_org_level_activity_logs) caused production incident with 2000+ lock contention sessions despite analyzer marking it as "✅ Safe - No locks". This was misleading - ALL ALTER TABLE operations acquire ACCESS EXCLUSIVE locks by default.

Changes:
- formatters.py: Changed SAFE description from "No locks" to "No contention risk"
- formatters.py: Updated NEEDS_REVIEW description to clarify brief lock requirement
- operations.py: Changed AddField nullable scoring from 0 to 1 (still SAFE level)
- operations.py: Updated reason text to mention brief lock requirement
- analyzer.py: Updated docstring with more accurate scoring descriptions
- test_risk_analyzer.py: Updated test expectation for nullable field score

Scoring remains:
- Score 0: Truly no contention (CreateModel, concurrent operations)
- Score 1: Brief lock on existing tables (still SAFE level, but honest about locks)
- Score 2-3: Extended operations
- Score 4-5: Table rewrites, breaking changes

webjunkie committed:
feat(ci): upload migration analysis as GitHub artifact

Add artifact upload step to preserve migration_analysis.md for debugging.
Uses if-no-files-found: ignore to avoid failing when no migrations present.

webjunkie committed:
fix(migration-analyzer): move score 1 operations to NEEDS_REVIEW category

Change RiskLevel ranges to accurately reflect lock impact:
- SAFE (score 0): Only truly no-contention operations (CreateModel, concurrent ops)
- NEEDS_REVIEW (score 1-3): Operations requiring locks or review
- BLOCKED (score 4-5): Unchanged

This fixes the inconsistency where AddField nullable (score 1) showed under SAFE section with 'No contention risk' header despite requiring ACCESS EXCLUSIVE lock.

Addresses PR review feedback that previous categorization perpetuated misleading messaging about locks.

webjunkie committed:
fix: add newline before closing backticks in missing migrations output

The code block was closing on the same line as the content, breaking markdown formatting in GitHub comments.

timgl commented on posthog/management/migration_analysis/operations.py:
```diff
@@ -45,12 +45,20 @@ def analyze(self, op) -> OperationRisk:
+
+For high-traffic tables, consider:
+- Deploy during low-traffic periods
+- Monitor lock contention and query timeouts during deployment
```
Maybe out of scope for this PR but could we be one step more specific on how to do this?

webjunkie committed:
feat(ci): update migration bot comments in-place with timestamps

Change both migration bot comments (SQL Changes and Risk Analysis) to update
in-place instead of deleting and recreating:
- Uses GitHub PATCH API to update existing comments
- Preserves comment position in PR thread and any replies/reactions
- Adds timestamp and commit SHA to footer for tracking
- Falls back to creating new comment if none exists
- Less visual churn in PR discussions

Format: *Last updated: YYYY-MM-DD HH:MM UTC (abc1234)*
